lib/Backend/GlobOptArrays.cpp

```diff
@@ -0,0 +1,1896 @@
+            {
+                return false;
+            }
+            // fall through
+
+        case Js::OpCode::LdElemI_A:
+        case Js::OpCode::LdMethodElem:
+            if (!instr->GetSrc1()->IsIndirOpnd())
+            {
+                return false;
+            }
+
+            baseOwnerIndir = instr->GetSrc1()->AsIndirOpnd();
+            baseOpnd = baseOwnerIndir->GetBaseOpnd();
+
+            // LdMethodElem is currently not profiled
+            isProfilableLdElem = instr->m_opcode != Js::OpCode::LdMethodElem;
+            needsBoundChecks = true;
+            needsHeadSegmentLength = true;
+            needsHeadSegment = true;
+            isLoad = true;
+            break;
+
+            // SIMD_JS
+        case Js::OpCode::Simd128_StArr_F4:
+        case Js::OpCode::Simd128_StArr_I4:
+            // no type-spec for Asm.js
+            if (globOpt->GetIsAsmJSFunc())
+            {
+                return false;
+            }
+            // fall through
+
+        case Js::OpCode::StElemI_A:
+        case Js::OpCode::StElemI_A_Strict:
+        case Js::OpCode::StElemC:
+            if (!instr->GetDst()->IsIndirOpnd())
+            {
+                return false;
+            }
+
+            baseOwnerIndir = instr->GetDst()->AsIndirOpnd();
+            baseOpnd = baseOwnerIndir->GetBaseOpnd();
+
+            isProfilableStElem = instr->m_opcode != Js::OpCode::StElemC;
+            needsBoundChecks = isProfilableStElem;
+            needsHeadSegmentLength = true;
+            needsHeadSegment = true;
+            isStore = true;
+            break;
+
+        case Js::OpCode::InlineArrayPush:
+        case Js::OpCode::InlineArrayPop:
+        {
+            IR::Opnd * thisOpnd = instr->GetSrc1();
+
+            // Abort if it not a LikelyArray or Object with Array - No point in doing array check elimination.
+            if (!thisOpnd->IsRegOpnd() || !thisOpnd->GetValueType().IsLikelyArrayOrObjectWithArray())
+            {
+                return false;
+            }
+
+            baseOwnerInstr = instr;
+            baseOpnd = thisOpnd->AsRegOpnd();
+
+            isLoad = instr->m_opcode == Js::OpCode::InlineArrayPop;
+            isStore = instr->m_opcode == Js::OpCode::InlineArrayPush;
+            needsLength = true;
+            needsHeadSegmentLength = true;
+            needsHeadSegment = true;
+            break;
+        }
+
+        case Js::OpCode::LdLen_A:
+            if (!instr->GetSrc1()->IsRegOpnd())
+            {
+                return false;
+            }
+
+            baseOpnd = instr->GetSrc1()->AsRegOpnd();
+            if (baseOpnd->GetValueType().IsLikelyObject() && baseOpnd->GetValueType().GetObjectType() == ObjectType::ObjectWithArray)
+            {
+                return false;
+            }
+
+            baseOwnerInstr = instr;
+            needsLength = true;
+            break;
+
+        default:
+            return false;
+    }
+
+    return true;
```

Comment:
>true [](start = 11, length = 4)

nit: should be false

Reply:
That last return is intentionally true. Any case that the original function would have returned early returns false, and true indicates that optimization can continue.

Reply:
Before refactoring we would have return from default so any case missed should still be false.

---
In reply to: [173604323](https://github.com/Microsoft/ChakraCore/pull/4805#discussion_r173604323) [](ancestors = 173604323)

Reply:
Sorry didn't see there are no returns in case arms.

---
In reply to: [173605004](https://github.com/Microsoft/ChakraCore/pull/4805#discussion_r173605004) [](ancestors = 173605004,173604323)